src/QsCompiler/Transformations/Attributes.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+
+
+namespace Microsoft.Quantum.QsCompiler.Transformations
+{
+    using CallablePredicate = Func<QsCallable, bool>;
+    using AttributeId = QsNullable<UserDefinedType>;
+    using QsRangeInfo = QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>;
+    using AttributeSelection = IEnumerable<(QsDeclarationAttribute, Func<QsCallable, bool>)>;
```

Comment:
If I understand right, the shared state here is a collection of filters that describe when to apply each new attribute to a given callable?

Reply:
Correct.